src/VisualStudio/Core/Def/Implementation/Experimentation/KeybindingResetDetector.cs

```diff
@@ -217,37 +261,87 @@ private void ShowGoldBar()
                               action: InfoBarClose));
         }
 
-        private ReSharperStatus IsReSharperEnabled()
+        /// <summary>
+        /// Returns true if ReSharper is installed, enabled, and not suspended.  
+        /// </summary>
+        private async ValueTask<ReSharperStatus> IsReSharperRunningAsync(ReSharperStatus lastStatus, CancellationToken cancellationToken)
         {
-            AssertIsForeground();
-
             // Quick exit if resharper is either uninstalled or not enabled
-            if (!_resharperExtensionEnabled)
+            if (!_resharperExtensionInstalledAndEnabled)
             {
                 return ReSharperStatus.NotInstalledOrDisabled;
             }
 
-            if (_oleCommandTarget == null)
-            {
-                _oleCommandTarget = _serviceProvider.GetService<IOleCommandTarget, SUIHostCommandDispatcher>();
-            }
+            await EnsureOleCommandTargetAsync().ConfigureAwait(false);
 
-            var cmds = new OLECMD[1];
-            cmds[0].cmdID = SuspendId;
-            cmds[0].cmdf = 0;
+            var cmds_suspend = new OLECMD[1];
+            cmds_suspend[0].cmdID = SuspendId;
+            cmds_suspend[0].cmdf = 0;
 
-            var hr = _oleCommandTarget.QueryStatus(ReSharperCommandGroup, (uint)cmds.Length, cmds, IntPtr.Zero);
-            if (ErrorHandler.Failed(hr))
+            var cmds_resume = new OLECMD[1];
+            cmds_resume[0].cmdID = ResumeId;
+            cmds_resume[0].cmdf = 0;
+
+            // poll until either suspend or resume botton is available, or until operation is canceled
+            while (true)
             {
+                cancellationToken.ThrowIfCancellationRequested();
+
+                var hr_suspend = await QueryStatusAsync(cmds_suspend).ConfigureAwait(false);
```

Comment:
💡 Consider the following for readability:

1. Rather than pass `OLECMD[]` to `QueryStatusAsync`, you can pass `int cmdId` (either `SuspendId` or `ResumeId`) and construct the `OLECMD[]` within that method.
2. Rather than query both commands and then process both commands, you can query one command and process it, and then only query the second command if the first did not yield an answer.

Reply:
Good idea.  I'll do this in a separate PR.

Reply:
Here is the follow-up issue: https://github.com/dotnet/roslyn/issues/32176